### ISSUE TITLE
Building awesome on Ubuntu 16

### ISSUE THREAD

**Issue by morethanoneanimal** [awesomeWM/awesome#1048]
I'm trying to build awesome and I'm following instructions on wiki. I have trouble running _make_ command, here's the output:

```
Running cmake…
-- git -> /usr/bin/git
-- asciidoc -> /usr/bin/asciidoc
-- xmlto -> /usr/bin/xmlto
-- gzip -> /bin/gzip
-- ldoc -> /usr/bin/ldoc
-- convert -> /usr/bin/convert
-- Checking for modules 'glib-2.0;gdk-pixbuf-2.0;cairo;x11;x11-xcb;xcb-cursor;xcb-randr;xcb-xtest;xcb-xinerama;xcb-shape;xcb-util>=0.3.8;xcb-keysyms>=0.3.4;xcb-icccm>=0.3.8;xcb-xkb;xkbcommon;xkbcommon-x11;cairo-xcb;libstartup-notification-1.0>=0.10;xproto>=7.0.15;libxdg-basedir>=1.0.0'
--   
--   
--   
CMake Error at /usr/share/cmake-3.5/Modules/FindPkgConfig.cmake:367 (message):
  A required package was not found
Call Stack (most recent call first):
  /usr/share/cmake-3.5/Modules/FindPkgConfig.cmake:532 (_pkg_check_modules_internal)
  awesomeConfig.cmake:112 (pkg_check_modules)
  CMakeLists.txt:18 (include)


-- Configuring incomplete, errors occurred!
See also "/home/moni/code/awesome/.build-kotel-i686-linux-gnu-5.4.0/CMakeFiles/CMakeOutput.log".
See also "/home/moni/code/awesome/.build-kotel-i686-linux-gnu-5.4.0/CMakeFiles/CMakeError.log".
Makefile:22: recipe for target 'cmake' failed
make: *** [cmake] Error 1
```

So required package is not found... but which one? How can I tell? Log files are completely useless and provide no information whatsoever, did anybody have similar issue / know what should I do?

**Comment by Elv13** [comment]
You can use  `sudo apt-get build-dep awesome` to do much of the work for you. After that it will only be missing the xcb keyboard library. But you are right, the error message is useless.

**Comment by blueyed** [comment]
Can we improve the usage of `pkg_check_modules` here?
But it looks like it should be just smarter by default in case of errors..

**Comment by morethanoneanimal** [comment]
With ```sudo apt-get build-dep awesome``` I'm getting error saying I should fill *sources.list* with some URIs, do you know which ones?
Indeed it looks like an issue with CMake. Some specific case maybe, because e.g. I got error for xcb-cursor missing and couple of others and once there were fixed I started getting just plain errors without explanation.

**Comment by Elv13** [comment]
> sources.list with some URIs, do you know which ones?

Uncomment all "deb-src" lines or whatever they are called. They are there by default, but disabled.

**Comment by morethanoneanimal** [comment]
Alright, that helped (for the record, ```sudo apt-get update``` after modifying *sources.list* and then I needed  ```sudo apt-get install lua-lgi-dev```.
So now that I have it installed... how do I run it? I tried putting ```exec awesome``` into my ~/.xinitrc but Unity started as usual. 
I recollect that when I installed awesome I was able to choose which environment to use, Unity or Awesome at login panel. I don't have this option now, is there some way to get that option?

**Comment by Elv13** [comment]
If Awesome is installed in /usr (using `cmake -DCMAKE_INSTALL_PREFIX=/usr (note that you have to clean your build directory first)) it should show up in the login menu. To use `.xinitrc`, you need to start X using `startx` instead of logind and ore lightdm. If installed correctly, there should be an `awesome.desktop` file in`/usr/share/xsessions/`

**Comment by blueyed** [comment]
You might also want to try the "make package" target we have to build a `.deb` that you can then install.
I would not recommend installing to `/usr` manually.

**Comment by morethanoneanimal** [comment]
I start to think I'm this guy: https://xkcd.com/1586/
I managed to install it via *.deb, the only thing that is still confusing me is missing rc.lua file, I can't find it neither in /etc/xdg nor ~/.config therefore I can't start it (but at least I got two options in my login panel).
I don't see any standalone rc.lua in sources, so any ideas how can I get this file?

**Comment by blueyed** [comment]
Yeah, maybe the deb does not contain it?!

**Comment by blueyed** [comment]
@morethanoneanimal 
You can use `dpkg --contents foo.deb` to list the contents of the generated `foo.deb`, and look for "etc" in there.

**Comment by psychon** [comment]
According to `ninja package && ar p awesome-3.5.2\~git1939-g68aae5f-dirty-Linux.deb data.tar.gz | gunzip - | tar --list | grep rc.lua`, there is the following file in the package: `/usr/local/etc/xdg/awesome/rc.lua`
I guess that awesome won't find the file there by default. We use the "xdg stuff" to find the config and that standard doesn't really expect something like this, I guess.

**Comment by morethanoneanimal** [comment]
On slightly different subject: what does your setup look like? What I'd expect is something like this: current, 'prod' version installed via package manager or built from master and a second one, developer/experimental (multiple instances possibly).
How do you quickly switch between them, how do you manage binaries?
How do I manage awesome install location? I did `make`, how do I specify destination of `make install`?

**Comment by blueyed** [comment]
@morethanoneanimal 
I am using a PKGBUILD on Arch Linux, and have the source linked therein.  I am using that package then for running awesome, and even when testing developing:

```sh
  cd ~/builds/awesome-git

  cur_version=$(pacman -Qi awesome-git|grep -E 'Version|Build Date')
  PKGEXT=.tar makepkg -srief
  new_version=$(pacman -Qi awesome-git|grep -E 'Version|Build Date')

  if [ "$new_version" = "$cur_version" ]; then
    echo "Version has not changed."
    exit 1
  fi
```

The return status is used to indicate if it should be restarted.

Others are using Xephyr, running from the development tree.
I have `run-from-git_pychon.sh` here from @psychon, but it might be out of date.
Would be nice to have this checked into Git.

Use `cd build; cmake .. -DCMAKE_INSTALL_PREFIX=/opt/awesome; make`.

I have a X session runner, via `~/.xinitrc` and `startx` that loops and restarts awesome (https://github.com/blueyed/dotfiles/blob/1aa03a780964d7d1ef204b9bc16e15ea048c164b/usr/bin/session-run-awesome.sh, but not up-to-date).  This allows to just kill awesome and then it restarts with the new binary.

**Comment by blueyed** [comment]
@morethanoneanimal 
It might be nice if you could come up with some README / documentation in this regard, collected from your findings and the input from here..!

**Comment by morethanoneanimal** [comment]
Sound like a good idea, it'll probably yet another 'Building awesome on Ubuntu' doc on [building awesome](https://awesome.naquadah.org/wiki/Building_awesome).
I'll mess around a little with Xephyr to add some info about that as well.

**Comment by morethanoneanimal** [comment]
I'm still a little confused on how to manage different versions. I have stable version installed via apt-get and I wanted to try newest version built from source on Xephyr. So I ran `cd build; cmake .. -DCMAKE_INSTALL_PREFIX=/opt/awesome; make`.
But at this point I still need to run `sudo make install` and this is the step I'm worried about, wouldn't it break my existing, stable version (e.g. by overwriting some files in /usr/local or somewhere else)?
I think there's no simple way to undo those changes (something like `make uninstall`), so probably I should use *.deb packages. AFAIK you can only have one version of given program at the time, so packet manager would remove my currently running package and install experimental one. That sounds dangerous too.
Is there a better way?

**Comment by blueyed** [comment]
> by overwriting some files in /usr/local or somewhere else

It will use `/opt/awesome` to install everything.
And you can have different versions being installed at the same time using different prefixes.
With a prefix like this, it's also easy to uninstall: just remove the whole directory.

> so probably I should use *.deb packages

Yeah, using packages is the best after all.

**Comment by morethanoneanimal** [comment]
Ok, let me present better one-liner here:
`cd build && cmake .. -DCMAKE_INSTALL_PREFIX=/opt/awesome && make`
I suggest updating previous answers with it, it may save some careless soul like mine in the future.

I ran the previous one without build directory so both `cd` and `cmake` failed but `make` succeeded and it ran without proper prefix set, but there was no clear indication of error. After that `sudo install` wreaked havoc in stable version files.

**Comment by morethanoneanimal** [comment]
Is there a separate step to update version tag? I'm building from master, last commit from 7th of November and `./awesome --version` returns `awesome v3.5.2-2174-g588cbc1 (The Fox).`.
My stable version is `awesome v3.5.9 (Mighty Ravendark)`. Counterintuitive.

**Comment by Elv13** [comment]
> Is there a separate step to update version stamp? 

Your not going to like my answer, but some help with https://github.com/awesomeWM/awesome/milestone/2 would "fix" your version tag issue by causing v4.0 (was to be v3.6) to be released. They are mostly documentation fixes and investigation. None of us have much time for these right now.

**Comment by morethanoneanimal** [comment]
To answer my question more precisely: version stamp is dictated by `git describe` and from `master` branch the closest tag is, indeed, v.3.5.2. Tag v3.5.9 can be found in branch `3.5`, which leads to conclusion that master branch and branch with current repository version are not merged.

**Comment by Elv13** [comment]
The version priority issue has been fixed with the release of 4.0.
